userdb/memory/mem.go

```diff
@@ -0,0 +1,96 @@
+package memory
+
+import (
+	"bufio"
+	"encoding/csv"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/janekolszak/idp/core"
+	"golang.org/x/crypto/bcrypt"
+)
+
+type Store struct {
+	hashes map[string]string
+	mtx    sync.RWMutex
+}
+
+func NewMemStore() (*Store, error) {
+	s := Store{}
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.hashes = make(map[string]string)
+
+	return &s, nil
+}
+
+func (s *Store) LoadHtpasswd(filename string) error {
+
+	f, err := os.OpenFile(filename, os.O_RDONLY, os.ModeExclusive)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	r := csv.NewReader(bufio.NewReader(f))
+	r.Comment = '#'
+	r.Comma = ':'
+	r.TrimLeadingSpace = true
+	r.FieldsPerRecord = 2
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.hashes = make(map[string]string)
+	for {
+		fields, err := r.Read()
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+
+			return err
+		}
+
+		s.hashes[fields[0]] = fields[1]
+	}
+}
+
+func (s *Store) Check(username, password string) error {
+
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	hash, exists := s.hashes[username]
+	// possibly compare against zero hash to prevent timing attack
+	err := bcrypt.CompareHashAndPassword([]byte(hash), []byte(password))
+	if !exists {
+		return core.ErrorNoSuchUser
```

Comment:
Yes, it's Provider's job to check the password. 

Storage's job is to store/retrive some credential data that will allow Provider to authenticate the user. Store doesn't know what kind of data it will be.